File: apps/quant-web/src/api/runtime.ts
import request from './request'
import type { RuntimeAlertProjection } from '../utils/runtimeHealthTypes'

export { normalizeRuntimeAlertProjection } from '../utils/runtimeHealthTypes'
export type { RuntimeAlertProjection, RuntimeAlertRuleErrorType, RuntimeAlertRuleStatus } from '../utils/runtimeHealthTypes'

export interface RuntimeComponentHealth {
  status: string
  latency_ms: number | null
  error_type: string | null
  error_message: string | null
}

export interface RuntimeLiveMarketHealth {
  status: string
  configured_enabled: boolean
  operational_count: number
  subscribed_count: number
  last_heartbeat_at: string | null
  last_bar_at: string | null
  phase_counts: Record<string, number>
  error_type: string | null
  error_message: string | null
}

export interface RuntimeAfterMarketFailureNotification {
  attempted_at: string
  state: 'provider_accepted' | 'failed'
  error_type: string | null
}

export interface RuntimeAfterMarketRun {
  trading_day: string
  status: string
  attempts: number | null
  started_at: string
  finished_at: string
  products: string[]
  error_code: string | null
  failure_notification: RuntimeAfterMarketFailureNotification | null
}

export interface RuntimeAfterMarketCurrentRun {
  scheduled_date: string
  started_at: string
  products: string[]
  attempt?: number
  stage?: string
  updated_at?: string
  stage_started_at?: string
  elapsed_seconds?: number
  current_symbol?: string | null
  current_partition?: { dataset: string[]; year: number; month: number } | null
  counters?: Record<string, { completed: number; total?: number }>
  stage_durations?: Record<string, number>
  retry_at?: string | null
}

export interface RuntimeAfterMarketHealth {
  status: string
  configured_enabled: boolean
  run_state: string
  expected_trading_day: string | null
  current_run: RuntimeAfterMarketCurrentRun | null
  last_run: RuntimeAfterMarketRun | null
  last_successful_trading_day: string | null
  last_failure: Record<string, string> | null
  error_type: string | null
  error_message: string | null
}

export interface RuntimeWeeklyAuditHealth {
  status: 'not_run' | 'running' | 'passed' | 'findings' | 'failed' | 'skipped_busy' | 'stuck' | 'stale' | 'invalid'
  readonly: true
  scope: 'operational_full_history'
  through: string | null
  finding_count: number | null
  started_at: string | null
  updated_at: string | null
  finished_at: string | null
  completed: number | null
  total: number | null
  current_symbol: string | null
}

export interface RuntimeAlertNotificationHealth {
  transport: string
  configured: boolean
  audience_count: number
  would_send: boolean
}

export interface RuntimeAlertHealth {
  status: string
  configured_enabled: boolean
  notification: RuntimeAlertNotificationHealth
  last_heartbeat_at: string | null
  enabled_rule_count: number
  scope_product_count: number
  processing_state: string
  notification_state: string
  last_processed_bar_at: string | null
  last_processing_success_at: string | null
  last_processing_failure_at: string | null
  processing_error_type: string | null
  last_event_at: string | null
  last_transport_attempt_at: string | null
  last_provider_accepted_at: string | null
  last_notification_failure_at: string | null
  notification_acknowledged_at: string | null
  notification_error_type: string | null
  consecutive_notification_failures: number
  error_type: string | null
  rule_status?: RuntimeAlertProjection['rule_status']
}

export interface RuntimeHealthResponse {
  status: string
  generated_at: string
  readonly: boolean
  would_start_services: boolean
  would_enqueue_jobs: boolean
  would_send_notifications: boolean
  components: {
    db: RuntimeComponentHealth
    redis: RuntimeComponentHealth
    live_market: RuntimeLiveMarketHealth
    after_market: RuntimeAfterMarketHealth
    weekly_audit?: RuntimeWeeklyAuditHealth
    alert: RuntimeAlertHealth
  }
}

export function getRuntimeHealth() {
  return request.get<never, RuntimeHealthResponse>('/api/runtime/health')
}

File: services/quant-api/app/schemas/runtime.py
"""Runtime 运维健康检查响应模型（Pydantic）。"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class RuntimeComponentHealth(BaseModel):
    """通用组件健康（DB、Redis 等单点探测）。"""

    status: str
    latency_ms: float | None = None
    error_type: str | None = None
    error_message: str | None = None


class RuntimeLiveMarketHealth(BaseModel):
    """Live V1 的短 TTL Redis heartbeat 快照。"""

    status: str
    configured_enabled: bool = False
    operational_count: int = 0
    subscribed_count: int = 0
    last_heartbeat_at: str | None = None
    last_bar_at: str | None = None
    phase_counts: dict[str, int] = Field(default_factory=dict)
    error_type: str | None = None
    error_message: str | None = None


class RuntimeAfterMarketFailureNotification(BaseModel):
    """Owner PushPlus 的一次 provider acceptance 尝试，不代表送达。"""

    attempted_at: str
    state: Literal["provider_accepted", "failed"]
    error_type: str | None = None


class RuntimeAfterMarketRun(BaseModel):
    """盘后一次运行的公开摘要。"""

    trading_day: str
    status: str
    attempts: int
    started_at: str
    finished_at: str
    products: list[str] = Field(default_factory=list)
    error_code: str | None = None
    failure_notification: RuntimeAfterMarketFailureNotification | None = None


class RuntimeAfterMarketCurrentRun(BaseModel):
    """当前盘后自然运行的 crash-visible 摘要。"""

    scheduled_date: str
    started_at: str
    products: list[str] = Field(default_factory=list)


class RuntimeAfterMarketHealth(BaseModel):
    """由本地公开状态文件派生的盘后维护摘要。"""

    status: str
    configured_enabled: bool = False
    run_state: str = "disabled"
    expected_trading_day: str | None = None
    current_run: RuntimeAfterMarketCurrentRun | None = None
    last_run: RuntimeAfterMarketRun | None = None
    last_successful_trading_day: str | None = None
    last_failure: dict[str, str] | None = None
    error_type: str | None = None
    error_message: str | None = None


class RuntimeAlertNotificationHealth(BaseModel):
    """Secret-safe structural notification audience readiness."""

    transport: str
    configured: bool = False
    audience_count: int = 0
    would_send: bool = False


class RuntimeAlertHealth(BaseModel):
    """Alert activation、notification transport 与短 TTL heartbeat 摘要。"""

    status: str
    configured_enabled: bool = False
    notification: RuntimeAlertNotificationHealth
    last_heartbeat_at: str | None = None
    enabled_rule_count: int = 0
    scope_product_count: int = 0
    processing_state: str = "unobserved"
    notification_state: str = "unobserved"
    last_processed_bar_at: str | None = None
    last_processing_success_at: str | None = None
    last_processing_failure_at: str | None = None
    processing_error_type: str | None = None
    last_event_at: str | None = None
    last_transport_attempt_at: str | None = None
    last_provider_accepted_at: str | None = None
    last_notification_failure_at: str | None = None
    notification_acknowledged_at: str | None = None
    notification_error_type: str | None = None
    consecutive_notification_failures: int = 0
    error_type: str | None = None


class RuntimeHealthComponents(BaseModel):
    """各子组件健康快照的容器。"""

    db: RuntimeComponentHealth
    redis: RuntimeComponentHealth
    live_market: RuntimeLiveMarketHealth
    after_market: RuntimeAfterMarketHealth
    alert: RuntimeAlertHealth


class RuntimeHealthOut(BaseModel):
    """``/api/runtime/health`` 顶层响应；顶层 readonly 标志声明无写操作授权。"""

    status: str
    generated_at: str
    readonly: bool = True
    would_start_services: bool = False
    would_enqueue_jobs: bool = False
    would_send_notifications: bool = False
    components: RuntimeHealthComponents
